partitura/musicanalysis/voice_separation.py

```diff
@@ -23,58 +28,51 @@ def rename_voices(voices):
                        dtype=voices.dtype)
 
 
-# def prepare_notearray(notearray):
-#     # * check whether notearray is a structured array
-#     # * check whether it has pitch/onset/duration fields
-#     # * return a copy of pitch/onset/duration fields with added id field
-#     if notearray.dtype.fields is None:
-#         raise ValueError('`notearray` must be a structured numpy array')
+def prepare_notearray(note_info):
+    # * return a copy of pitch/onset/duration fields with added id field
 
-#     fields = ('pitch', 'onset', 'duration')
-#     time_units = ('div', 'quarter', 'beat', 'sec')
-#     onset_field = None
-#     duration_field = None
+    # get note array from note info
+    notearray = ensure_notearray(note_info)
 
-#     for unit in time_units:
-#         onset_field = f'onset_{unit}'
-#         if onset_field in notearray.dtype.names:
-#             break
-#     for unit in time_units:
-#         duration_field = f'duration_{unit}'
-#         if duration_field in notearray.dtype.names:
-#             break
-#     # if field not in notearray.dtype.names:
-#     #     raise ValueError('Input array does not contain required field {0}'.format(field))
+    # Get onset and duration units
+    onset_unit, duration_unit = get_time_units_from_note_array(notearray)
 
-#     dtypes = dict(notearray.dtype.descr)
-#     new_dtype = [(n, dtypes[m]) for n, m in zip(fields, ['pitch', onset_field, duration_field])] + [('id', 'i4')]
+    req_fields = ('pitch', onset_unit, duration_unit)
+    # new array will be unit agnostic
+    new_fields = ('pitch', 'onset', 'duration')
+    for field in req_fields:
+        if field not in notearray.dtype.names:
+            raise ValueError('Input array does not contain required field {0}'.format(field))
 
+    dtypes = dict(notearray.dtype.descr)
+    new_dtype = ([(nn, dtypes[n]) for nn, n in zip(new_fields, req_fields)] +
+                 [('id', 'i4')])
 
-#     return np.fromiter(zip(notearray['pitch'],
-#                            notearray['onset_div'],
-#                            notearray['duration_div'],
-#                            np.arange(len(notearray))),
-#                        dtype=new_dtype)
+    return np.fromiter(zip(notearray['pitch'],
+                           notearray[onset_unit],
+                           notearray[duration_unit],
+                           np.arange(len(notearray))),
+                       dtype=new_dtype)
 
 
 def argmax_pitch(idx, pitches):
     return idx[np.argmax(pitches[idx])]
 
 
-def estimate_voices(notearray, monophonic_voices=False):
-    """Voice estimation using the voice separation algorithm proposed 
-    in [1]_.
+def estimate_voices(note_info, monophonic_voices=False):
+    """Voice estimation using the voice separation algorithm
+       proposed in [1]_.
```

Comment:
Could you use the number 6 for the reference? That way there are no collisions with the references when the docstrings of all the musicanalysis functions are aggregated into a single html file.

Reply:
Fixed in b9a7917.